### PR TITLE
Changed optimizer to optax

### PR DESCRIPTION
**Pull request by berndbohnet**
# What does this PR do?
This PR changes the optimzer optax as requested. 

I tested the code again on local gpu and same result as before. 

Fixes # (issue)

## Checklist
- [x ] This PR fixes a minor issue (e.g.: typo or small bug) or improves the docs (you can dismiss the other
      checks if that's the case).
- [] This change is discussed in a Github issue/
      [discussion](https://github.com/google/flax/discussions) (please add a
      link).
- [ ] The documentation and docstrings adhere to the
      [documentation guidelines](https://github.com/google/flax/blob/main/docs/README.md#how-to-write-code-documentation).
- [ ] This change includes necessary high-coverage tests.
      (No quality testing = no merge!)

### CODE REVIEW

File: examples/nlp_seq/train.py [hunk 1]
```diff
@@ -191,35 +192,38 @@ def compute_metrics(logits, labels, weights):
...
+               learning_rate_fn,
+               dropout_rng=None):
   """Perform a single training step."""
-
   train_keys = ['inputs', 'targets']
   (inputs, targets) = [batch.get(k, None) for k in train_keys]
 
-  weights = jnp.where(targets > 0, 1, 0).astype(jnp.float32)
-  dropout_rng, new_dropout_rng = random.split(dropout_rng)
+  weights = jnp.where(inputs > 0, 1, 0).astype(jnp.float32)
```
Comment (marcvanzee) [inline examples/nlp_seq/train.py:204]: Nit: is this change from `targets` to `inputs` intended?

**Review by marcvanzee — APPROVED** [review]
Looks great, thanks a lot! Thanks also for the other cleanups/improvements you did!

File: examples/nlp_seq/train.py (hunk 1 above)
Comment (berndbohnet) [reply examples/nlp_seq/train.py:204]: Very well spotted. No, this was not intended - makes not difference as both are padding. However, let's move back to targets. I will update.

Comment (marcvanzee) [reply examples/nlp_seq/train.py]: Thanks!

**Comment by andsteing** [comment]
Part of #1053 and #568

File: examples/nlp_seq/train.py [hunk 2]
```diff
@@ -26,8 +26,9 @@
 from absl import logging
 from flax import jax_utils
 from flax import linen as nn
+from flax.training import train_state
 from flax.metrics import tensorboard
-from flax import optim
+import optax
```
Comment (andsteing) [inline examples/nlp_seq/train.py:31]: nit: consider sorting imports alphabetically

File: examples/nlp_seq/train.py [hunk 3]
```diff
@@ -191,35 +192,38 @@ def compute_metrics(logits, labels, weights):
...
+    """loss function used for training."""
     logits = model.apply({'params': params}, inputs=inputs, train=True,
                          rngs={'dropout': dropout_rng})
     loss, weight_sum = compute_weighted_cross_entropy(logits, targets, weights)
+
     mean_loss = loss / weight_sum
     return mean_loss, logits
 
-  step = optimizer.state.step
+  step = state.step
```
Comment (andsteing) [inline examples/nlp_seq/train.py:217]: nit: is this variable needed? it's only used once below (and once above where you refer to it by `state.step`)

File: examples/nlp_seq/train.py [hunk 4]
```diff
@@ -326,7 +342,8 @@ def eval_step(params, batch):
   for step, batch in zip(range(num_train_steps), train_iter):
     batch = common_utils.shard(jax.tree_map(lambda x: x._numpy(), batch))  # pylint: disable=protected-access
 
-    optimizer, metrics, dropout_rngs = p_train_step(optimizer, batch, dropout_rng=dropout_rngs)
+    state, metrics = p_train_step(
```
Comment (andsteing) [inline examples/nlp_seq/train.py:345]: nit: would this fit on a single line?

**Review by andsteing — COMMENTED** [review]
The change looks good, I just added some small nits.

Did you run the example to verify the same results are produced? In that case adding a link to the TensorBoard.dev trace would be great (and maybe even adding/updating the link in the README.md)

File: examples/nlp_seq/train.py (hunk 2 above)
Comment (berndbohnet) [reply examples/nlp_seq/train.py:31]: Thanks. done.

File: examples/nlp_seq/train.py (hunk 3 above)
Comment (berndbohnet) [reply examples/nlp_seq/train.py:217]: No. Changed it.

File: examples/nlp_seq/train.py (hunk 4 above)
Comment (berndbohnet) [reply examples/nlp_seq/train.py:345]: Yes, fits. Changed.

**Comment by berndbohnet** [comment]
please, have another look. The commits got a bit strange names by accident but the formatting is now included.

**Comment by berndbohnet** [comment]
Thanks for the comment.  Yes. I am running it now again and I do the updates when finished in few hours.  Accuracy seems alright a bit higher with the same settings for the optimizer, maybe previous changes caused this too.

File: examples/nlp_seq/train.py (hunk 2 above)
Comment (andsteing) [reply examples/nlp_seq/train.py:31]: did you push the changes?

File: examples/nlp_seq/train.py (hunk 3 above)
Comment (andsteing) [reply examples/nlp_seq/train.py:217]: did you push the changes?

File: examples/nlp_seq/train.py (hunk 4 above)
Comment (andsteing) [reply examples/nlp_seq/train.py:345]: did you push the changes?

File: examples/nlp_seq/README.md [hunk 5]
```diff
@@ -21,7 +21,7 @@ The model should run with other configurations and hardware, but explicitly test
 
 | Hardware |  Batch size  | Learning rate | Training time | Accuracy  | TensorBoard.dev |
 |:---:|:---:|:---:|:---:|:---:|:---:|
-| Nvidia V100 (16GB) | 64  |  0.05 | 5h 15m | 72.20% | [2020-03-22](https://tensorboard.dev/experiment/YkUAdwYaQ9OtYl2IVe3MvA/) |
+| Nvidia Titan V (12GB) | 64  |  0.05 | 41m | 92.5% | [2022-02-18](https://tensorboard.dev/experiment/dZTE3FoYSxarf29dKji10w/) |
```
Comment (andsteing) [inline examples/nlp_seq/README.md:24]: The traces look markedly different:
1. `accuracy/xpos_train` was ~1.00 previously, now is ~0.1
2. `accuracy/xpos_eval` was ~0.7 previously, now is ~0.9
3. `best_dev_score` was ~0.72 previously, now is ~0.92
4. `accuracy` measurements have much less variance
5. `best_dev_score` was fewer "jumps" than before

What happened here?

**Review by andsteing — CHANGES_REQUESTED** [review]

File: examples/nlp_seq/README.md [hunk 6]
```diff
@@ -21,7 +21,7 @@ The model should run with other configurations and hardware, but explicitly test
 
 | Hardware |  Batch size  | Learning rate | Training time | Accuracy  | TensorBoard.dev |
 |:---:|:---:|:---:|:---:|:---:|:---:|
-| Nvidia V100 (16GB) | 64  |  0.05 | 5h 15m | 72.20% | [2020-03-22](https://tensorboard.dev/experiment/YkUAdwYaQ9OtYl2IVe3MvA/) |
+| Nvidia Titan V (12GB) | 64  |  0.05 | 1:36h | 68.6% | [2022-04-29](https://tensorboard.dev/experiment/F5ULHlyzQlieVJn5PG8mRQ/) |
```
Comment (andsteing) [inline examples/nlp_seq/README.md:24]: hm... the linked tensorboard shows `xpos_eval=0.6928` after `5h58m21s` ...

how come this decreased so much from the previous 72.20% ?

File: examples/nlp_seq/train.py (hunk 2 above)
Comment (andsteing) [reply examples/nlp_seq/train.py:31]: I still see

```python
from flax.training import train_state
from flax.metrics import tensorboard
```

sorting alphabetically, it should instead be 
```python
from flax.metrics import tensorboard
from flax.training import train_state
```

File: examples/nlp_seq/train.py [hunk 7]
```diff
@@ -296,17 +300,29 @@ def initialize_variables(init_rng):
...
-
   learning_rate_fn = create_learning_rate_scheduler(
       base_learning_rate=learning_rate)
 
+  optimizer = optax.adamw(
+      learning_rate_fn, b1=0.9, b2=0.98, eps=1e-9,
+      weight_decay=1e-1
+      )
+  state = train_state.TrainState.create(
+    apply_fn=model.apply,
```
Comment (andsteing) [inline examples/nlp_seq/train.py:311]: nit: we usually indent with 4 spaces for function args

File: examples/nlp_seq/train.py [hunk 8]
```diff
@@ -296,17 +300,29 @@ def initialize_variables(init_rng):
...
-  optimizer = optimizer_def.create(init_variables['params'])
-  optimizer = jax_utils.replicate(optimizer)
-
   learning_rate_fn = create_learning_rate_scheduler(
       base_learning_rate=learning_rate)
 
+  optimizer = optax.adamw(
+      learning_rate_fn, b1=0.9, b2=0.98, eps=1e-9,
+      weight_decay=1e-1
+      )
```
Comment (andsteing) [inline examples/nlp_seq/train.py:309]: nit: we usually move the closing parens on the line before

**Review by andsteing — CHANGES_REQUESTED** [review]
There are still a couple of open comments. Thanks for addressing / resolving them.

**Comment by andsteing** [comment]
One more thing: in the end you'll probably need to squash some of the commits because we request to only have 5 commits per PR.
https://github.com/google/flax/blob/625dc1664c9fc98aaf1b5226ca8f57f322fb59a5/.github/workflows/build.yml#L35

**Comment by berndbohnet** [comment]
Thanks for the comments. I tried to squash it. I hope that worked. 

I tried to figure out what causes the accuracy change. I could not spot it. I suggest to leave it for now and later to improve it. The example is only educational and does not aim for accuracy.

File: examples/nlp_seq/README.md (hunk 6 above)
Comment (berndbohnet) [reply examples/nlp_seq/README.md]: I do not know what cause this. I have read the code again and as it is only an education example, I would suggest to go ahead. We might investigate later again.

File: examples/nlp_seq/train.py (hunk 7 above)
Comment (berndbohnet) [reply examples/nlp_seq/train.py]: done. Thanks.

File: examples/nlp_seq/train.py (hunk 8 above)
Comment (berndbohnet) [reply examples/nlp_seq/train.py]: done. Thanks

File: examples/nlp_seq/train.py [hunk 9]
```diff
@@ -55,7 +55,7 @@
     help=('Frequency of eval during training, e.g. every 1000 steps.'))
 
 flags.DEFINE_integer(
-    'num_train_steps', default=75000, help=('Number of train steps.'))
+    'num_train_steps', default=20000, help=('Number of train steps.'))
```
[the remark this answers is not in the document]
Comment (berndbohnet) [reply examples/nlp_seq/train.py]: Changed back to 75k.

File: examples/nlp_seq/README.md [hunk 10]
```diff
@@ -21,7 +21,7 @@ The model should run with other configurations and hardware, but explicitly test
 
 | Hardware |  Batch size  | Learning rate | Training time | Accuracy  | TensorBoard.dev |
 |:---:|:---:|:---:|:---:|:---:|:---:|
-| Nvidia V100 (16GB) | 64  |  0.05 | 5h 15m | 72.20% | [2020-03-22](https://tensorboard.dev/experiment/YkUAdwYaQ9OtYl2IVe3MvA/) |
+| Nvidia Titan V (12GB) | 64  |  0.05 | 1:36h | 92.5% | [2022-04-29](https://tensorboard.dev/experiment/AuVcQWAGSnqdRMc9kekeog) |
```
[the remark this answers is not in the document]
Comment (berndbohnet) [reply examples/nlp_seq/README.md]: See comments below. Issue remains.

[the remark this answers is not in the document]
Comment (berndbohnet) [reply examples/nlp_seq/README.md]: I agree training accuracy should be close to 100%. Appart from that I would not report it as it should be around 98%.

**Comment by berndbohnet** [comment]
alright. I did a rebase, I hope this resolves it.

**Review by andsteing — CHANGES_REQUESTED** [review]
you'll need to cleanup the 16 commits to a maximum of 5 (maybe 1 is enough) before we can merge the PR

**Comment by berndbohnet** [comment]
I think, this should be fine now?

File: examples/nlp_seq/train.py (hunk 2 above)
Comment (andsteing) [reply examples/nlp_seq/train.py]: Reordered.

File: examples/nlp_seq/train.py (hunk 4 above)
Comment (andsteing) [reply examples/nlp_seq/train.py]: Yes we did.

File: examples/nlp_seq/README.md (hunk 5 above)
Comment (andsteing) [reply examples/nlp_seq/README.md]: Will be examined separately in #2201

File: examples/nlp_seq/README.md (hunk 10 above)
[the remark this answers is not in the document]
Comment (andsteing) [reply examples/nlp_seq/README.md]: Will be resolved in #2201

**Review by andsteing — APPROVED** [review]
Now it looks all good, thanks!
